platform/c2000/inverter_drivers/pmicdriver.h

```diff
@@ -0,0 +1,433 @@
+    { TLF35584_PROTCFG, TLF35584_PROTCFG_UNLOCK_KEY3 },
+    { TLF35584_PROTCFG, TLF35584_PROTCFG_UNLOCK_KEY4 },
+    { TLF35584_WDCFG1, TLF35584_WDCFG1_WDSLPEN | TLF35584_WDCFG1_FWDETHR(14) },
+    { TLF35584_WDCFG0,
+      TLF35584_WDCFG0_WWDETHR(14) | TLF35584_WDCFG0_WWDEN |
+          TLF35584_WDCFG0_FWDEN | TLF35584_WDCFG0_WWDTSEL |
+          TLF35584_WDCFG0_WDCYC_1MS },
+    { TLF35584_SYSPCFG1,
+      TLF35584_SYSPCFG1_SS2DEL_0MS | TLF35584_SYSPCFG1_ERRREC_1MS },
+    { TLF35584_FWDCFG, TLF35584_FWDCFG_WDHBTP_CYCLES(250) },
+    { TLF35584_WWDCFG0, TLF35584_WWDCFG0_CW_CYCLES(50) },
+    { TLF35584_WWDCFG1, TLF35584_WWDCFG1_OW_CYCLES(100) },
+    { TLF35584_PROTCFG, TLF35584_PROTCFG_LOCK_KEY1 },
+    { TLF35584_PROTCFG, TLF35584_PROTCFG_LOCK_KEY2 },
+    { TLF35584_PROTCFG, TLF35584_PROTCFG_LOCK_KEY3 },
+    { TLF35584_PROTCFG, TLF35584_PROTCFG_LOCK_KEY4 }
+};
+
+static const uint16_t RegisterConfigSize =
+    sizeof(RegisterConfig) / sizeof(RegisterConfig[0]);
+
+/**
+ * Functional Watchdog response sequence structure
+ */
+struct WatchdogResponse
+{
+    uint16_t resp3;
+    uint16_t resp2;
+    uint16_t resp1;
+    uint16_t resp0;
+};
+
+/**
+ * Canned Functional Watchdog response sequences from Table 26 in Section 15.3
+ * functional Watchdog of the TLF35584 datasheet
+ */
+static const WatchdogResponse WatchdogResponses[0x10] = {
+    // clang-format off
+    { 0xFF, 0x0F, 0xF0, 0x00 },
+    { 0xB0, 0x40, 0xBF, 0x4F },
+    { 0xE9, 0x19, 0xE6, 0x16 },
+    { 0xA6, 0x56, 0xA9, 0x59 },
+    { 0x75, 0x85, 0x7A, 0x8A },
+    { 0x3A, 0xCA, 0x35, 0xC5 },
+    { 0x63, 0x93, 0x6C, 0x9C },
+    { 0x2C, 0xDC, 0x23, 0xD3 },
+    { 0xD2, 0x22, 0xDD, 0x2D },
+    { 0x9D, 0x6D, 0x92, 0x62 },
+    { 0xC4, 0x34, 0xCB, 0x3B },
+    { 0x8B, 0x7B, 0x84, 0x74 },
+    { 0x58, 0xA8, 0x57, 0xA7 },
+    { 0x17, 0xE7, 0x18, 0xE8 },
+    { 0x4E, 0xBE, 0x41, 0xB1 },
+    { 0x01, 0xF1, 0x0E, 0xFE },
+    // clang-format on
+};
+
+static const uint16_t WatchdogResponsesSize =
+    sizeof(WatchdogResponses) / sizeof(WatchdogResponses[0]);
+
+static const int StateTransitionDelay = 100; // uS
+
+
+/**
+ * \brief Verify that function x didn't fail. Return immediately if it did.
+ * Assumes a local variable "result" of type TeslaM3PowerWatchdog::Error
+ */
+#define CHECK(x)                                                               \
+    if ((result = x) != OK)                                             \
+    {                                                                          \
+        return result;                                                         \
+    }
+
+/**
+ * \brief Define a scoped SPI transaction that manually asserts the ~CS line for
+ * the duration of a SPI transaction.
+ */
+struct SPITransaction
+{
+    SPITransaction()
+    {
+        // gpio_clear(GPIO_PMIC_CS_BANK, GPIO_PMIC_CS);
+    }
+
+    ~SPITransaction()
+    {
+        // gpio_set(GPIO_PMIC_CS_BANK, GPIO_PMIC_CS);
+    }
+};
+
+
+
+template <typename SpiDriverT>
```

Comment:
This looks like a good approach to isolating the HW dependencies and allowing code to be unit tested. I had wanted to avoid using too many templates because it might intimidate contributors from the wider openinverter user base. The improvements in readability and testability are more than worth it so go for it!